[webdriver] Move Webdriver to ServoShell

longvatrong111 opened the pull request:
Moving `webdriver` to `servoshell`:

- Let `servoshell` manage lifecycle of `webdriver`
- One by one, move the handling of webdriver commands from `constellation` to `embedder`

Partially fix: https://github.com/servo/servo/issues/37370
Partially fix webdriver test timeout with `no_top_browsing_context`

cc: @xiaochengh

yezhizhen commented on components/script/script_thread.rs:
```diff
@@ -1055,6 +1064,12 @@ impl ScriptThread {
         }
     }
 
+    fn process_pending_script_events(&self) {
```
```suggestion
    fn process_pending_webdriver_script_events(&self) {
```

yezhizhen commented on components/script/script_thread.rs:
```diff
@@ -1278,6 +1293,7 @@ impl ScriptThread {
             // https://drafts.csswg.org/css-position-4/#process-top-layer-removals.
         }
 
+        self.process_pending_script_events();
```
```suggestion
        self.process_pending_webdriver_script_events();
```

yezhizhen commented on components/script/script_thread.rs:
```diff
@@ -2089,6 +2105,39 @@ impl ScriptThread {
+        // `self.documents`, which would conflict with the immutable borrow of it that
+        // occurs for the rest of the messages
+        match msg {
+            WebDriverScriptCommand::ExecuteScript(ref _script, ref _reply) => {
+                return self.pending_webdriver_script_events.borrow_mut().push(
+                    WebdriverScriptEvent {
+                        pipeline_id,
+                        msg,
+                        can_gc,
+                    },
+                );
+            },
+            WebDriverScriptCommand::ExecuteAsyncScript(ref _script, ref _reply) => {
+                return self.pending_webdriver_script_events.borrow_mut().push(
+                    WebdriverScriptEvent {
+                        pipeline_id,
+                        msg,
+                        can_gc,
+                    },
+                );
+            },
```
Since they are same, you can merge into one.
```suggestion
            WebDriverScriptCommand::ExecuteScript(..) | WebDriverScriptCommand::ExecuteAsyncScript(..) => {
                return self.pending_webdriver_script_events.borrow_mut().push(
                    WebdriverScriptEvent {
                        pipeline_id,
                        msg,
                        can_gc,
                    },
                );
            },
```

longvatrong111 replied:
Thanks. Let me update.

yezhizhen reviewed (commented):
Thanks for working on this!

longvatrong111 commented:
> Does that mean that every embedder will need to add Webdriver support code?

It's a nice question. I think we can separate Webdriver support code as much as possible and reuse in different embedders.

xiaochengh commented:
Discussed with @longvatrong111 offline.

It's better to make WebDriver independent of ServoShell, as WebDriver shouldn't depend on any specific embedder.

WebDriver and real events should still share the same code paths as much as possible.

mrobinson commented:
> It's better to make WebDriver independent of ServoShell, as WebDriver shouldn't depend on any specific embedder.

I still think it isn't a good idea to make it totally independent. At the very least, servoshell should use some generic WebDriver interface, but still execute the requested actions using its internal interface. WebDriver should be driving a browser, and not just the embedding layer.

yezhizhen commented:
As discussed with @longvatrong111 , I will move change in `script_thread.rs` and contribute upon it.

xiaochengh commented:
> WebDriver should be driving a browser, and not just the embedding layer.

+1 to this. We'll see how it will be carried out.

yezhizhen commented:
It seems new assertion failure is introduced for `tests\wpt\tests\webdriver\tests\classic\element_click\events.py`, with changes in `script_thread.rs` here.
```
assertion failed: !GetCurrentRealmOrNull(*cx).is_null() (thread Script(1,2), at components\script_bindings\realms.rs:25)
```
I will start by working from there.

longvatrong111 commented:
Rebase PR. Will check all reviews 1 by 1. 
Right now I'm focusing on fixing synchronization issue between `Input Event` and `WebDriver Script Event` with this PR.

xiaochengh commented:
Is there a clearer illustration / doc of what the relation between servoshell and web driver is, and what the relation should be?

mrobinson commented:
> Is there a clearer illustration / doc of what the relation between servoshell and web driver is, and what the relation should be?

Yes, sorry for the confusion.

The WebDriver is an HTTP server that can drive the browser via commands from tests that are sent to the server.

**Currently**: The WebDriver is activated by the Servo embedding API. When active, the Constellation spawns a thread to listen for HTTP requests and send IPC messages directly to the Constellation or ScriptThreads. This presents an issue for several reasons including:

1.  When a client requests a new WebView to be created, the embedding API must request the embedder to create the WebView. It makes more sense for the WebDriver to create the WebView through the embedding API directly. This allows testing the embedding API and prevents a weird inversion of control. The only other time this happens is via things like `window.open()`, but these are more closely managed by the ScriptThread.
2. Some events must be processed by the compositor before being sent to the internals of Servo. Since the WebDriver is running inside Servo, these events must be sent "out" of Servo to the embedding layer and it is difficult to ensure proper ordering of event handling since some events go straight to the Constellation and some go back out to the embedding layer first.

**Future**: servoshell would run a WebDriver thread that runs the HTTP server. Many commands become Servo embedding API commands and a good amount of special code can be removed from Servo to deal with the WebDriver. Note that this will still require an API to send some of the more specific WebDriver commands to the Constellation and then to the ScriptThread. This means that things like opening and closing WebViews work exactly the same way they do in a browser application. In order to avoid having the WebDriver code depend directly on servoshell, servoshell may be able to provide some callbacks which the WebDriver code can use.

longvatrong111 committed:
Move webdriver to servoshell

Signed-off-by: batu_hoang <longvatrong111@gmail.com>

longvatrong111 commented:
@xiaochengh @mrobinson it is confirmed that the implementation commands in [Context section](https://www.w3.org/TR/webdriver2/#contexts) are much simpler if sending events to servoshell instead of constellation. For example [Close Window](https://www.w3.org/TR/webdriver2/#close-window)

yezhizhen reviewed (commented):
> it is confirmed that the implementation of event route of commands in [Context section](https://www.w3.org/TR/webdriver2/#contexts) are much simpler if sending events to servoshell instead of constellation. For example [Close Window](https://www.w3.org/TR/webdriver2/#close-window)

I don't think so. Right now close window shares `Constellation::handle_close_top_level_browsing_context` with request `ScriptToConstellationMessage` and `EmbedderToConstellationMessage::CloseWebView`. Isn't that nice?


> One by one, move the handling of webdriver commands from constellation to embedder

Some of them would still be easier to handle in Constellation as they require information from it. Need to be very careful with each case... :(

longvatrong111 commented:
The only kind of issues we need to deal is synchronization between webdriver commands, some go to embedder, some go to constellation first.
For worst case scenario, we can add some wait.

xiaochengh commented:
> The only kind of issues we need to deal is synchronization between webdriver commands, some go to embedder, some go to constellation first. For worst case scenario, we can add some wait.

Could you give an example?

Since WebDriver mimics a user, I thought all commands should go through embedder first?

longvatrong111 commented on ports/servoshell/desktop/app.rs:
```diff
@@ -296,6 +309,41 @@ impl App {
+            return;
+        };
+
+        match webdriver_receiver.try_recv() {
```
I agree, look like a big TODO note.

mrobinson commented:
> Anyway, I agree that we can send all through embedder first for synchronization purpose. It may help later.

I think for these type of commands we can add a new embedder API just for WebDriver implementations called `Servo::execute_webdriver_command` or `WebView::execute_webdriver_command`. This would be reserved just for the commands that need direct access to the DOM.

xiaochengh commented:
> > Anyway, I agree that we can send all through embedder first for synchronization purpose. It may help later.
> 
> I think for these type of commands we can add a new embedder API just for WebDriver implementations called `Servo::execute_webdriver_command` or `WebView::execute_webdriver_command`. This would be reserved just for the commands that need direct access to the DOM.

+1.

This also uses the same route as our ArkWeb embedder, which also has commands to interact directly with the script threads.

xiaochengh reviewed (commented):
Thanks for the improvements!

I'm also wondering if we can further improve it by decoupling the current implementation with desktop ServoShell, so that it becomes something that can be easily plugged into any embedder implementation (eg SimpleServo on OHOS, which might be of our interests).

Probably something like:

```rust
struct WebDriverEmbedderPlugin {
  webdriver_receiver: Receiver<EmbedderMsg>,
  embedder_proxy: EmbedderProxy,
  ...
}

impl WebDriverEmbedderPlugin<'app> {
  // Called from embedder to start a WebDriver server.
  pub fn new(...) -> Self {...}

  // Called from embedder only when running.
  pub fn handle_messages(&mut self, running_state: Rc<RunningAppState>) {...}

  fn forward_command(&self, command) {... }
}

...

pub struct App {
  ...
  webdriver_plugin: Option<WebDriverEmbedderPlugin>,
  ...
}

```

But there is no need to do it in this PR, since there are quite some obstacles before there, like we need a common trait for desktop and egl `AppState`, and probably more. But it's worth leaving a TODO for it.

xiaochengh reviewed (approved):
And the current PR LGTM as is. But let's make sure @mrobinson also takes a look.

yezhizhen commented on ports/servoshell/desktop/app.rs:
```diff
@@ -435,6 +509,9 @@ impl ApplicationHandler<WakerEvent> for App {
         // Consume and handle any events from the Minibrowser.
         self.handle_servoshell_ui_events();
```
I suggest we can disable this for webdriver test.
```suggestion
    if self.opts.webdriver_port.is_none(){
        self.handle_servoshell_ui_events();
    }
```

yezhizhen reviewed (commented):
The code itself looks good. Very excited for this!

---------
Side notes:
I've been verifying the execution path, and found some redundancy for channel communication which has been there for quite a while. (Not really related to this PR)

But I wanna ask: why we sometimes use `crossbeam_channel`, and sometimes `ipc`? I assume the former is more efficient + mpmc, the latter is required for inter-process communication (to Content Process, e.g. script thread). If I'm correct, then we are misusing many ipc channels in `webdriver_server/lib.rs` if only talking to Constellation/embedder. Instead, should just use `crossbeam_channel` or even `mpsc`?
cc @jdm @sagudev

mrobinson commented:
Apologies for not getting to this one yet. I will look at it first thing tomorrow morning.

mrobinson commented on ports/servoshell/desktop/app.rs:
```diff
@@ -154,6 +157,18 @@ impl App {
         let servo = servo_builder.build();
         servo.setup_logging();
 
+        // Init webdriver server here before servo is moved
```
```suggestion
        // Initialize WebDriver server here before `servo` is moved.
```

mrobinson commented on ports/servoshell/desktop/app.rs:
```diff
@@ -154,6 +157,18 @@ impl App {
+                webdriver_server::create_embedder_channel(self.waker.clone());
+            self.webdriver_receiver = Some(receiver);
+
+            // TODO: Remove this channel from webdriver once
+            // we move all webdriver events to embedder
```
```suggestion
            // TODO: WebDriver will no longer need this channel once all WebDriver
            // commands are executed via the Servo API.
```

mrobinson commented on components/constellation/constellation.rs:
```diff
@@ -4699,8 +4699,8 @@ where
-            WebDriverCommandMsg::CloseWebView(webview_id) => {
-                self.handle_close_top_level_browsing_context(webview_id);
+            WebDriverCommandMsg::CloseWebView(..) => {
+                unreachable!("Moved to embedder");
```
```suggestion
                unreachable!("This command should be send directly to the embedder.");
```

mrobinson commented on components/constellation/constellation.rs:
```diff
@@ -4738,9 +4738,8 @@ where
-                let is_open = self.webviews.get(webview_id).is_some();
-                let _ = response_sender.send(is_open);
+            WebDriverCommandMsg::IsWebViewOpen(..) => {
+                unreachable!("Moved to embedder");
```
```suggestion
                unreachable!("This command should be sent directly to the embedder.");
```

mrobinson commented on components/webdriver_server/lib.rs:
```diff
@@ -122,8 +122,25 @@ fn cookie_msg_to_cookie(cookie: cookie::Cookie) -> Cookie {
+
+pub fn start_server(
+    port: u16,
+    constellation_chan: Sender<EmbedderToConstellationMessage>,
```
Let's have this channel only accept `WebDriverCommandMsg` instead of adding a new variant to `EmbedderMsg`. `EmbedderMsg` should be used for messages that come from the "inside" of Servo out to the embedder. Only sending `WebDriverCommandMsg` on this channel should simplify the code.

```suggestion
    constellation_chan: Sender<EmbedderToConstellationMessage>,
```

This will ensure that various kinds of weird situations are automatically impossible (such as receiving other kinds of `EmbedderMsg`s on this channel).

longvatrong111 replied:
This channel exists a long time ago, and provided by `constellation`. We can not change it because all components use this to send msg to `constellation`.
Fortunately, we can remove this channel from `webdriver` when all msg are sent to `embedder`.
Maybe you confuse this one with `embedder_sender`?

xiaochengh replied:
As a trick, we can name it `constellation_chan_deprecated` to indicate that this is something we want to remove from WebDriver.

mrobinson commented on components/webdriver_server/lib.rs:
```diff
@@ -122,8 +122,25 @@ fn cookie_msg_to_cookie(cookie: cookie::Cookie) -> Cookie {
 
-pub fn start_server(port: u16, constellation_chan: Sender<EmbedderToConstellationMessage>) {
-    let handler = Handler::new(constellation_chan);
+pub fn create_embedder_channel(
+    event_loop_waker: Box<dyn EventLoopWaker>,
+) -> (EmbedderProxy, Receiver<EmbedderMsg>) {
+    let (sender, receiver) = unbounded();
+    (
+        EmbedderProxy {
+            sender,
+            event_loop_waker,
+        },
+        receiver,
+    )
+}
```
I think you could just store the `event_loop_waker` and `Sender<...>` separately in the `Handler` and then have a method called `send_message_to_embedder` that is responsible for triggering the `event_loop_waker`.

longvatrong111 replied:
Thank you, nice change, it makes sending message much simpler.

mrobinson commented on components/webdriver_server/lib.rs:
```diff
@@ -213,10 +230,15 @@ struct Handler {
 
     session: Option<WebDriverSession>,
 
+    /// Embedder proxy
+    embedder_proxy: EmbedderProxy,
+
```
If you follow the suggestion above:

```suggestion
    /// A [`Sender`] that sends messages to the embedder that this `WebDriver instance controls.
    /// In addition to sending a message, we must always wake up the embedder's event loop so it
    /// knows that more messages are available for processing.
    embedder_sender: Sender<WebDriverCommandMsg>,

    /// An [`EventLoopWaker`] which is used to wake up the embedder event loop.
    event_loop_waker: Box<dyn EventLoopWaker>,
```

mrobinson reviewed (changes requested):
Great stuff!

I just have a few suggestions.

longvatrong111 committed:
Decompose webdriver::embedder_proxy

Decompose webdriver::embedder_proxy to a sender and a waker.
Make the sender channel uses only webdriver command message.
Rebase main and solve conflict

Signed-off-by: batu_hoang <longvatrong111@gmail.com>

longvatrong111 committed:
Rename constellation_chan as a TODO note

Signed-off-by: batu_hoang <longvatrong111@gmail.com>

mrobinson commented on ports/servoshell/desktop/app.rs:
```diff
@@ -48,6 +49,7 @@ pub struct App {
     t_start: Instant,
     t: Instant,
     state: AppState,
+    webdriver_receiver: Option<Receiver<WebDriverCommandMsg>>,
```
Apologies, I missed one comment. Please move this to `RunningAppState`, which will make it more obvious what its lifetime is.

mrobinson commented on ports/servoshell/desktop/app.rs:
```diff
@@ -154,6 +157,22 @@ impl App {
         servo.setup_logging();
 
+        // Initialize WebDriver server here before `servo` is moved.
+        if let Some(port) = self.opts.webdriver_port {
```
Then this can just be:

```suggestion
        let webdriver_receiver = self.opts.webdriver_port.map(|port| {
            let (embedder_sender, embedder_receiver) = unbounded();

            // TODO: WebDriver will no longer need this channel once all WebDriver
            // commands are executed via the Servo API.
            let constellation_sender_deprecated = servo.constellation_sender();
            webdriver_server::start_server(
                port,
                constellation_sender_deprecated,
                embedder_sender,
                self.waker.clone(),
            );
            
            embedder_receiver
        });
```

mrobinson reviewed (approved):
Thanks! One more comment before landing, but once that's done feel free to land without another review.

longvatrong111 committed:
Move webdriver_receiver to app running state

Signed-off-by: batu_hoang <longvatrong111@gmail.com>

mrobinson commented:
I added some minor rustdoc comments and did a little re-ordering of members. Sending to the merge queue now. Thanks @longvatrong111!
